### DISCUSSION TITLE
Winpython 3.11.4 sqlalchemy inspector.get_table_names(schema=schema) not working after upgrading to 2+ from 1.4.48

### DISCUSSION THREAD

**Discussion by ceaza** [sqlalchemy/sqlalchemy#10183] [Usage Questions] [answered]
inspector.get_table_names(schema=schema) is no longer working on MSSQL. Am I using the method correctly? See code snippet below:

```
connection_string = "mssql+pyodbc:///?odbc_connect=%s" % connection_string
engine = db.create_engine(connection_string) #, echo=True)
inspector = db.inspect(engine)
inspector.get_table_names(schema=schema)
```

Thanks

**Comment by CaselIT** [comment]
Hi,

From memory I don't think much has changed from v1.4 to v2 there.
What's the issue in the specific? were some tables listed in 1.4 and are no longer there in v2?

**Reply by ceaza** [reply]
Hi @CaselIT, this is the error returned, which is fatal.

Python/WPy64-31140/python-3.11.4.amd64/Lib/site-packages/sqlalchemy/engine/default.py?line=920
mexecute(statement, parameters)\n\n\u001b[1;31mDBAPIError\u001b[0m: (pyodbc.Error) 

[Microsoft][ODBC SQL Server Driver]Invalid precision value (0) (SQLBindParameter)'
[SQL: SELECT [INFORMATION_SCHEMA].[TABLES].[TABLE_NAME] 
FROM [INFORMATION_SCHEMA].[TABLES] 
WHERE [INFORMATION_SCHEMA].[TABLES].[TABLE_SCHEMA] = CAST(? AS NVARCHAR(max)) AND [INFORMATION_SCHEMA].[TABLES].[TABLE_TYPE] = CAST(? AS NVARCHAR(max)) ORDER BY [INFORMATION_SCHEMA].[TABLES].[TABLE_NAME]]
[parameters: ('dw', 'BASE TABLE')]
(Background on this error at: https://sqlalche.me/e/20/dbapi)

I did updated the client driver but had no impact.

**Reply by ceaza** [reply]
Based on this issue, https://github.com/sqlalchemy/sqlalchemy/issues/8681 and adding  "use_setinputsizes=False" to create_engine resolved the issue. Apparently the main branch deals with the issue without the need of adding use_setinputsizes.

**Reply by zzzeek** [reply]
use_setinputsizes should never have any effect whatsoever on reflection queries, which are only selecting for simple string values, not numeric values.   If that is occurring, that's a major bug that would need to be fixed.     We would need to see the CREATE TABLE statement for this table that you are reflecting as well as information on encodings in use etc.

**Reply by ceaza** [reply]
@zzzeek it's a system table, so if I understand you correctly Microsoft are responsible for the CREATE TABLE statement.

**Reply by zzzeek** [reply]
oh it's just get_table_names() .  OK what is winpython ?  can you try this program from linux or osx to the same database so we can figure out if this is a client or server issue?

**Reply by ceaza** [reply]
Winpython https://winpython.github.io/ is a windows data science distribution for Python. Will take me a little while to get a linux Python version querying MSSQL.

**Reply by CaselIT** [reply]
what schema are you trying to reflect?

**Reply by ceaza** [reply]
DW

**Reply by CaselIT** [reply]
without more information about mssql version used (the dw schema seems to be of some azure data warehouse thing) and a reproducible script it's not really easy to make other suppositions

@gordthompson do you have ideas here?

**Reply by ceaza** [reply]
The version is SQL Server 15.0.2080.9. Its not on azure.

**Reply by CaselIT** [reply]
I have no dw schema on a default docker image for mssql https://hub.docker.com/_/microsoft-mssql-server

Does the error only happens on that schema or it's independent of the actual schema requested?

**Reply by ceaza** [reply]
DW is a user created schema. As far as I can see its totally independent.
This works:
schemas = inspector.get_schema_names()

But I am not able to get the table names for any of the schemas for the list returned. For example this fails for all schemas:

```
schema = 'sys'
for table_name in inspector.get_table_names(schema=schema):
    print(table_name)
```
As I have said, when I change the engine to engine = db.create_engine(connection_string,use_setinputsizes=False) it works.

**Reply by CaselIT** [reply]
I can't reproduce
The script below works for both 15 and 16 using pyodbc 4.0.35 and 4.0.39
```py
import sqlalchemy as sa

e = sa.create_engine('mssql+pyodbc://scott:tiger^5HHH@127.0.0.1:1433/test?driver=ODBC+Driver+18+for+SQL+Server&Encrypt=Optional')

with e.connect() as c:
    print(c.exec_driver_sql('select @@version').scalar())
i = sa.inspect(e)
schema = 'sys'
print(i.get_table_names(schema=schema))
```

this prints
```
Microsoft SQL Server 2022 (RTM-CU5) (KB5026806) - 16.0.4045.3 (X64)
        May 26 2023 12:52:08
        Copyright (C) 2022 Microsoft Corporation
        Developer Edition (64-bit) on Linux (Ubuntu 20.04.6 LTS) <X64>
[]
```
```
Microsoft SQL Server 2019 (RTM-CU21) (KB5025808) - 15.0.4316.3 (X64)
        Jun  1 2023 16:32:31
        Copyright (C) 2019 Microsoft Corporation
        Developer Edition (64-bit) on Linux (Ubuntu 20.04.6 LTS) <X64>
[]
```

what's your pyodbc version and odbc driver version? also what's your client os?

**Reply by ceaza** [reply]
I am on Windows 11, 

pyodbc.version is 4.0.39

`Microsoft SQL Server 2019 (RTM-GDR) (KB4583458) - 15.0.2080.9 (X64) 
	Nov  6 2020 16:50:01 
	Copyright (C) 2019 Microsoft Corporation
	Standard Edition (64-bit) on Windows Server 2019 Datacenter 10.0 <X64> (Build 17763: ) (Hypervisor)`

![image](https://github.com/sqlalchemy/sqlalchemy/assets/9589657/ab338c44-2304-48f4-9e42-4f4f4c34f15c)

**Reply by gordthompson** [reply]
What do you see if you

```
print(engine.raw_connection().getinfo(6))
```

**Reply by ceaza** [reply]
SQLSRV32.DLL

**Reply by gordthompson** [reply]
Yeah, so that's the ancient "SQL Server" driver I mentioned in my answer here

https://github.com/sqlalchemy/sqlalchemy/discussions/10183#discussioncomment-6645386

Your screenshot shows that you also have "ODBC Driver 18 for SQL Server" installed, so you should be using that one.

**Reply by ceaza** [reply]
Thanks @gordthompson. Do I remove it? As mentioned earlier in this post that I upgrade the driver using msodbcsql.msi. Same as the link you provided, but seems did not have an impact on SQLSRV32.dll.

**Reply by CaselIT** [reply]
No, you need to specify the driver name in he connection string like in my example https://github.com/sqlalchemy/sqlalchemy/discussions/10183#discussioncomment-6644974

**Reply by gordthompson** [reply]
They are totally separate drivers. Have another look at your screenshot. "ODBC Driver 18 for SQL Server" is MSODBCSQL18.DLL.

Your connection string should use `Driver=ODBC Driver 18 for SQL Server`.

**Comment by gordthompson** [comment, marked as answer]
This is a known issue with the ancient (and long-deprecated) "SQL Server" ODBC driver that ships with Windows. Use a current ODBC driver like "ODBC Driver 17 for SQL Server" or "ODBC Driver 18 for SQL Server".

https://learn.microsoft.com/en-us/sql/connect/odbc/download-odbc-driver-for-sql-server?view=sql-server-ver16

**Reply by ceaza** [reply]
Thanks so much for all the help!

Got the connection working with this:
```import pyodbc
import sqlalchemy as sa
server = 'server' 
driver = 'ODBC Driver 18 for SQL Server'
database = 'database' 

connection_string = f"DRIVER={driver};SERVER={server};DATABASE={database};Trusted_Connection=Yes;TrustServerCertificate=Yes"

connection_url = sa.engine.URL.create(
    "mssql+pyodbc", query={"odbc_connect": connection_string})

engine = sa.create_engine(connection_url)

with engine.connect() as c:
    print(c.exec_driver_sql('select @@version').scalar())
i = sa.inspect(engine)
schema = 'dw'
print(i.get_table_names(schema=schema))

pyodbc.version
print(engine.raw_connection().getinfo(6))

```

And output:
```
Microsoft SQL Server 2019 (RTM-GDR) (KB4583458) - 15.0.2080.9 (X64) 
	Nov  6 2020 16:50:01 
	Copyright (C) 2019 Microsoft Corporation
	Standard Edition (64-bit) on Windows Server 2019 Datacenter 10.0 <X64> (Build 17763: ) (Hypervisor)

[tables listed correctly]
msodbcsql18.dll
```
